File: daily/2024/1395_Count_Number_of_Teams.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <map>

using namespace std;

class BIT {
public:
    BIT(vector<int>& nums){
        n = nums.size();
        tree = vector<int>(n+1, 0);
        for (int i = 0; i < n; i++) update(i, nums[i]);
    }
    void update(int index, int value){
        index++;
        while(index <= n){
            tree[index] += value;
            index += index & -index;
        }
    }
    int query(int index){
        index++;
        int ans = 0;
        while(index > 0){
            ans += tree[index];
            index -= index & -index;
        }
        return ans;
    }
    int range_query(int left, int right) {return query(right) - query(left - 1);}
private:
    int n;
    vector<int> tree;
};

class Solution {
public:
    int numTeams(vector<int>& rating){
        int n = rating.size();
        int M = *max_element(rating.begin(), rating.end());
        vector<int> count(M+1, 0);
        BIT left_BIT(count);
        BIT right_BIT(count);
        for (auto i : rating) right_BIT.update(i, 1);
        int ans = 0;
        for (int middle_index = 0; middle_index < n; middle_index++){
            int num = rating[middle_index];
            right_BIT.update(num, -1);
            int left_smaller = left_BIT.range_query(1, num - 1);
            // int left_larger = left_BIT.range_query(num + 1, M);
            int left_larger = middle_index - left_smaller;
            int right_smaller = right_BIT.range_query(1, num - 1);
            // int right_larger = right_BIT.range_query(num + 1, M);
            int right_larger = n - 1 - middle_index - right_smaller;
            ans += left_smaller * right_larger + left_larger * right_smaller;
            left_BIT.update(num, 1);
        }
        return ans;
    }
};

// class Solution {
// public:
//     int numTeams(vector<int>& rating) {
//         int ans = 0;
//         ans += count(rating);
//         reverse(rating.begin(), rating.end());
//         ans += count(rating);
//         return ans;
//     }
//     int count(vector<int>& rating) {
//         int ans = 0;
//         map<int,int> m1;
//         map<int,int> m2;
//         for (int num : rating){
//             // cout << num << " : " << endl;
//             for (auto i : m2){
//                 if (i.first >= num) break;
//                 ans += i.second;
//             }
//             m1[num] = 1;
//             int temp = 0;
//             for (auto i : m1){
//                 if (i.first >= num) break;
//                 temp += i.second;
//             }
//             m2[num] = temp;
//             // cout << "ans = " << ans << endl;
//         }
//         return ans;
//     }
// };

int main(){
    vector<int> rating = {2,5,3,4,1};
    // vector<int> rating = {2,1,3};
    // vector<int> rating = {1,2,3,4};
    Solution S;
    cout << S.numTeams(rating) << endl;
    return 0;
}

File: daily_python/1395_Count_Number_of_Teams.py
import sys
import os
from typing import List

class Solution:
    def numTeams(self, rating: List[int]) -> int:
        max_element = max(rating)
        left_BIT = [0] * (max_element+1)
        right_BIT = [0] * (max_element+1)
        for num in rating:
            self.update(right_BIT, num, 1)
        ans = 0
        for num in rating:
            self.update(right_BIT, num, -1)
            left_smaller = self.query_range(left_BIT, 1, num-1)
            left_larger = self.query_range(left_BIT, num+1, max_element)
            right_smaller = self.query_range(right_BIT, 1, num-1)
            right_larger = self.query_range(right_BIT, num+1, max_element)
            ans += left_smaller * right_larger + left_larger * right_smaller
            self.update(left_BIT, num, 1)
        return ans
    
    def update(self, BIT: List[int], index: int, val: int) -> None:
        while(index < len(BIT)):
            BIT[index] += val
            index += index & (-index)
    def query(self, BIT: List[int], index: int)-> int:
        ans = 0
        while(index):
            ans += BIT[index]
            index -= index & (-index)
        return ans
    def query_range(self, BIT: List[int], left: int, right: int) -> int:
        return self.query(BIT, right) - self.query(BIT, left-1)

def readDataInFile(filename: str):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read()
        lines = content.split('\n')
        for data in lines:
            data = list(map(int,(data.split('=')[1].strip().strip('{};').split(','))))
            dataset.append(data)
    return dataset

if __name__ == "__main__":
    if (len(sys.argv) == 1):
        path = os.path.basename(__file__)
        filename = os.path.splitext(path)[0] + '.txt'
    else:
        filename = sys.argv[1]
    
    dataset = readDataInFile(filename)
    # print(dataset)
    solution = Solution()
    results = []
    for data in dataset:
        results.append(solution.numTeams(data))

    for index, result in enumerate(results):
        print(f'dataset {index+1} : {result}')